### PR TITLE
Fixed #24146 -- Allowed model._meta.get_field() to be used after apps.models_ready

### PR DESCRIPTION
**Pull request by PirosB3**

### CODE REVIEW

**Comment by charettes** [comment]
If I understand correctly this still allows forward fields to be retrieved before `models_ready`?

**Comment by aaugustin** [comment]
This is probably a good idea.

However, it means that, if models or fields are changed in `AppConfig.ready()`, some caches must be invalidated. Should we document this?

**Comment by carljm** [comment]
Yes, I was thinking about the same thing. I think a note in the documentation for `AppConfig.ready` explaining how to invalidate the appropriate cache(s?) if you add fields to models would make sense.

**Comment by PirosB3** [comment]
Thanks for the comments.  We get back with answers as soon as I'm home

**Comment by PirosB3** [comment]
Heard @aaugustin on IRC:

Between `models_ready` and `ready` there should not be any modifications to models, and relations between models. The only thing that happens in between these calls is an `AppConfig.ready()` signal for each model in the Apps registry.

Given that the public Meta API is read-only (modifying models is not supported and still isn't) this change should be safe. The only concern Aymeric expressed is that third-party apps could potentially modify models during the `AppConfig.ready()` stage (for instance a pluggable app could add created_at and updated_at fields to every model).

Personally, I don't see this as a blocker. If third-party app developers want to do these changes, I assume they have enough knowledge of the codebase to call `apps.clear_cache()` once they are done.

**Comment by carljm** [comment]
Hi PirosB3 - I think this is a good change. Since I'd committed to offering a PR for this ticket, just checking in to see if there's anything I can do to help?

**Comment by PirosB3** [comment]
Hi @carljm That's fine I think I am okay on this.
@aaugustin and anyone else involved: are we happy to proceed this way? If so, I'll add a few tests to make the changes stronger

**Comment by timgraham** [comment]
Yes, please add the additional tests.

**Comment by PirosB3** [comment]
ok @timgraham added tests that failed prior to the patch. Please let me know if this PR needs anything else, if not I'll squash commits

**Comment by PirosB3** [comment]
buildbot, retest this please.

**Comment by timgraham** [comment]
I think the `admin_checks` app is more appropriate place to add the test. Ideally, we can reuse or slightly alter existing models rather than adding new ones (since more models means slower tests).

**Comment by PirosB3** [comment]
@timgraham refactored

File: tests/admin_checks/tests.py [hunk 1]
```diff
@@ -667,3 +667,20 @@ class MyModelAdmin(admin.ModelAdmin):
             )
         ]
         self.assertEqual(errors, expected)
+
+    def test_list_filter_works_on_through_field_even_when_apps_not_ready(self):
+        """
+        Ensure list_filter can access reverse fields even when the Apps registry
+        is not ready.
```
Comment (timgraham) [inline tests/admin_checks/tests.py]: To save a line: "is not ready; refs ##24146."

File: tests/admin_checks/tests.py [hunk 2]
```diff
@@ -667,3 +667,20 @@ class MyModelAdmin(admin.ModelAdmin):
             )
         ]
         self.assertEqual(errors, expected)
+
+    def test_list_filter_works_on_through_field_even_when_apps_not_ready(self):
+        """
+        Ensure list_filter can access reverse fields even when the Apps registry
```
Comment (timgraham) [inline tests/admin_checks/tests.py]: It's the "app registry" (no 's')

File: tests/admin_checks/tests.py [hunk 3]
```diff
@@ -667,3 +667,20 @@ class MyModelAdmin(admin.ModelAdmin):
...
+        is not ready.
+        Refs #24146
+        """
+        class BookAdminWithListFilter(admin.ModelAdmin):
+            list_filter = ['authorsbooks__featured']
+
+        # Temporarily pretending apps are not ready yet. This issue can happen
+        # if the value of 'list_filter' refers to a 'through__field'.
+        Book._meta.apps.ready = False
+        errors = BookAdminWithListFilter.check(model=Book)
```
Comment (timgraham) [inline tests/admin_checks/tests.py]: to be safe:
```
try:
    errors = BookAdminWithListFilter.check(model=Book)
finally:
    Book._meta.apps.ready = True
```

**Comment by timgraham** [comment]
Ready for squashing, I'd say.

**Comment by timgraham** [comment]
merged in 19188826b4aa989475418f1ea9bf8631b04da1e8, thanks.

**Comment by collinanderson** [comment]
Works well. Thanks so much.

**Pull request closed without merging**